File: androidApp/src/main/java/com/habileducation/themovie/android/ui/movieDetail/MovieDetailScreen.kt
package com.habileducation.themovie.android.ui.movieDetail

import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.hilt.navigation.compose.hiltViewModel
import com.habileducation.themovie.android.R
import com.habileducation.themovie.android.ui.components.ErrorScreen
import com.habileducation.themovie.android.ui.components.ScreenContainer
import com.habileducation.themovie.viewState.MovieDetailViewState

/**
 * Created by Annas Surdyanto on 18/11/21.
 *
 */
@Composable
fun MovieDetailScreen(
    movieId: Long,
    viewModel: MovieDetailViewModel = hiltViewModel()
) {
    val state = viewModel.state.collectAsState().value
    if (state == MovieDetailViewState.empty) viewModel.initState(movieId)
    val title = state.movieResponse?.movieDetail?.title ?: "Movie Detail"
    ScreenContainer(
        loadingState = state.isLoading,
        onRefresh = { viewModel.initState(movieId) },
        barTitle = title,
        trailingIcon = if (state.movieResponse == null) null else if (state.isFavorited) R.drawable.favorited else R.drawable.unfavorited,
        onTrailingIconClicked = viewModel::setFavorite
    ) {
        state.movieResponse?.let {
            MovieDetailScreenContent(
                data = it,
            )
        }
        state.error?.let {
            ErrorScreen(message = it.message.toString()) {
                viewModel.initState(movieId)
            }
        }
    }
}

File: iosApp/iosApp/UI/MovieDetail/MovieDetailScreen.swift
//
//  MovieDetailScreen.swift
//  iosApp
//
//  Created by Developer on 29/11/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import SwiftUI
import shared

struct MovieDetailScreen: View {
    var movieId: Int64
    @ObservedObject var viewModel = MovieDetailViewModel()
    
    init(movieId: Int64){
        self.movieId = movieId
        viewModel.initState(movieId: movieId)
    }
    
    var body: some View {
        let state = viewModel.state
        
        //NavigationView{
            
                if (state.isLoading){ ProgressView() }
                
                if (state.movieResponse != nil) {
                    MovieDetailScreenContent(data: state.movieResponse!, onRefresh: {viewModel.initState(movieId: movieId)})
                }
                
                if (state.error != nil){
                    ErrorScreen(message: state.error!.message!){ viewModel.initState(movieId: movieId) }
                }
            
            //.navigationBarTitleDisplayMode(.inline)
        //}
    }
}

struct MovieDetailScreen_Previews: PreviewProvider {
    static var previews: some View {
        MovieDetailScreen(movieId: 24332)
    }
}
